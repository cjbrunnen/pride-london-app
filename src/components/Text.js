// @flow
import React from "react";
import { Text as RnText, StyleSheet } from "react-native";
import Markdown from "react-native-easy-markdown";
import type { Node } from "react";
import type { StyleObj } from "react-native/Libraries/StyleSheet/StyleSheetTypes";

export type TextType =
  | "h1"
  | "h2"
  | "h3"
  | "h4"
  | "h5"
  | "h6"
  | "text"
  | "small"
  | "xSmall";

type Props = {
  children: Node,
  type?: TextType,
  markdown?: boolean,
  style?: StyleObj
};

const Text = ({ children, type, markdown, style }: Props) =>
  markdown ? (
    <Markdown
      style={style}
      markdownStyles={{ ...textStyles, ...markdownStyles }}
    >
      {children}
    </Markdown>
  ) : (
    <RnText style={[type && styles[type], style]}>{children}</RnText>
  );
Text.defaultProps = {
  type: "text",
  markdown: false,
  style: {}
};

const textStyles = {
  h1: {
    fontFamily: "Poppins-Bold",
    fontSize: 24,
    lineHeight: 28
  },
  h2: {
    fontFamily: "Poppins-SemiBold",
    fontSize: 18,
    lineHeight: 24
  },
  h3: {
    fontFamily: "Poppins-SemiBold",
    fontSize: 16,
    lineHeight: 24
  },
  h4: {
    fontFamily: "Roboto-Medium",
    fontSize: 16,
    lineHeight: 24
  },
  h5: {
    fontFamily: "Roboto",
    fontSize: 16,
    lineHeight: 24
  },
  h6: {
    fontFamily: "Roboto",
    fontSize: 16,
    lineHeight: 24
  },
  text: {
    fontFamily: "Roboto",
    fontSize: 16,
    lineHeight: 24
  },
  small: {
    fontFamily: "Roboto",
    fontSize: 14,
    lineHeight: 20
  },
  xSmall: {
    fontFamily: "Roboto",
    fontSize: 12,
    lineHeight: 16
  }
};

const markdownStyles = {
  // "u" - underline is listed in react-native-easy-markdown
  // but doesn't exist in the markdown spec so should be rendered bold
  u: {
    fontWeight: "bold"
  }
};

const styles = StyleSheet.create(textStyles);

export default Text;
