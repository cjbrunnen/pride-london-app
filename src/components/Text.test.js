// @flow
import React from "react";
import { shallow } from "enzyme";
import Text from "./Text";

it("renders correctly", () => {
  const output = shallow(<Text>Some text</Text>);
  expect(output).toMatchSnapshot();
});

it("renders text based on type", () => {
  const output = shallow(<Text type="h1">Some text</Text>);

  expect(output.props().style[0]).toEqual(
    expect.objectContaining({ fontFamily: "Poppins-Bold" })
  );
});

it("renders markdown component based on flag", () => {
  const output = shallow(<Text markdown>**Some markdown**</Text>);

  expect(output).toMatchSnapshot();
});

it("passes custom styles down to component", () => {
  const style = { fontSize: 300 };
  const output = shallow(<Text style={style}>Some text</Text>);

  expect(output.props().style[output.props().style.length - 1]).toBe(style);
});
