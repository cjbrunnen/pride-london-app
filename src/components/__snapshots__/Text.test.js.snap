// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "fontFamily": "Roboto",
        "fontSize": 16,
        "lineHeight": 24,
      },
      Object {},
    ]
  }
>
  Some text
</Text>
`;

exports[`renders markdown component based on flag 1`] = `
<Markdown
  debug={false}
  markdownStyles={
    Object {
      "h1": Object {
        "fontFamily": "Poppins-Bold",
        "fontSize": 24,
        "lineHeight": 28,
      },
      "h2": Object {
        "fontFamily": "Poppins-SemiBold",
        "fontSize": 18,
        "lineHeight": 24,
      },
      "h3": Object {
        "fontFamily": "Poppins-SemiBold",
        "fontSize": 16,
        "lineHeight": 24,
      },
      "h4": Object {
        "fontFamily": "Roboto-Medium",
        "fontSize": 16,
        "lineHeight": 24,
      },
      "h5": Object {
        "fontFamily": "Roboto",
        "fontSize": 16,
        "lineHeight": 24,
      },
      "h6": Object {
        "fontFamily": "Roboto",
        "fontSize": 16,
        "lineHeight": 24,
      },
      "small": Object {
        "fontFamily": "Roboto",
        "fontSize": 14,
        "lineHeight": 20,
      },
      "text": Object {
        "fontFamily": "Roboto",
        "fontSize": 16,
        "lineHeight": 24,
      },
      "u": Object {
        "fontWeight": "bold",
      },
      "xSmall": Object {
        "fontFamily": "Roboto",
        "fontSize": 12,
        "lineHeight": 16,
      },
    }
  }
  parseInline={false}
  style={Object {}}
  useDefaultStyles={true}
>
  **Some markdown**
</Markdown>
`;
